loadtest/loadtest.go

```diff
@@ -75,26 +77,45 @@ func (lt *LoadTester) AddUsers(numUsers int) (int, error) {
 // addUser is an internal API called from Run and AddUsers both.
 // DO NOT call this by itself, because this method is not protected by a mutex.
 func (lt *LoadTester) addUser() error {
-	activeUsers := len(lt.controllers)
+	activeUsers := len(lt.activeControllers)
 	if activeUsers == lt.config.UsersConfiguration.MaxActiveUsers {
 		return ErrMaxUsersReached
 	}
-	userId := activeUsers + 1
+
+	controllerId := activeUsers + len(lt.idleControllers) + 1
+
+	userId := controllerId
 	// If specified by the config, we randomly pick an existing user again,
 	// to simulate multiple sessions.
 	if activeUsers != 0 && rand.Int()%lt.config.UsersConfiguration.AvgSessionsPerUser != 0 {
 		userId = rand.Intn(activeUsers)
 	}
-	controller, err := lt.newController(userId, lt.statusChan)
-	if err != nil {
-		return fmt.Errorf("could not create controller: %w", err)
+
+	var controller control.UserController
+
+	for id := range lt.idleControllers {
+		controllerId = id
+		controller = lt.idleControllers[id]
+		delete(lt.idleControllers, id)
+		break
+	}
+
+	if controller == nil {
+		var err error
+		controller, err = lt.newController(userId, lt.statusChan)
+		if err != nil {
+			return err
+		}
 	}
+
 	if err := controller.SetRate(lt.config.UserControllerConfiguration.Rate); err != nil {
 		return fmt.Errorf("loadtest: failed to set controller rate %w", err)
 	}
-	lt.controllers = append(lt.controllers, controller)
+
 	lt.status.NumUsers++
 	lt.status.NumUsersAdded++
+	lt.activeControllers[controllerId] = controller
```

Comment:
I'd be extra cautious and check whether we are not overwriting an existing active controller.